File: apps/packages/ui/src/services/quizzes.ts
import { bgRequest } from "@/services/background-proxy"
import type { AllowedPath } from "@/services/tldw/openapi-guard"
import { createResourceClient } from "@/services/resource-client"
import type {
  DeckSchedulerSettingsEnvelope,
  DeckSchedulerType,
  DeckReviewPromptSide,
  StudyAssistantContextResponse,
  StudyAssistantRespondRequest,
  StudyAssistantRespondResponse
} from "@/services/flashcards"

const quizzesClient = createResourceClient({
  basePath: "/api/v1/quizzes" as AllowedPath
})

export const QUIZ_GENERATION_TIMEOUT_MS = 120000

const quizAttemptsClient = createResourceClient({
  basePath: "/api/v1/quizzes/attempts" as AllowedPath,
  updateMethod: "PUT"
})

const getQuestionsClient = (quizId: number) =>
  createResourceClient({
    basePath: `/api/v1/quizzes/${quizId}/questions` as AllowedPath
  })

const getQuizAttemptsClient = (quizId: number) =>
  createResourceClient({
    basePath: `/api/v1/quizzes/${quizId}/attempts` as AllowedPath
  })

// Question types
export type QuestionType = "multiple_choice" | "multi_select" | "matching" | "true_false" | "fill_blank"
export type AnswerValue = number | string | number[] | Record<string, string>
export type QuizGenerateSourceType =
  | "media"
  | "note"
  | "flashcard_deck"
  | "flashcard_card"
  | "quiz_attempt"
  | "quiz_attempt_question"
export type QuizGenerateSource = {
  source_type: QuizGenerateSourceType
  source_id: string
}
export type SourceCitation = {
  source_type?: QuizGenerateSourceType | null
  source_id?: string | null
  label?: string | null
  quote?: string | null
  media_id?: number | null
  chunk_id?: string | null
  timestamp_seconds?: number | null
  source_url?: string | null
}

// Quiz container
export type Quiz = {
  id: number
  name: string
  description?: string | null
  workspace_id?: string | null
  workspace_tag?: string | null
  media_id?: number | null
  source_bundle_json?: QuizGenerateSource[] | null
  total_questions: number
  time_limit_seconds?: number | null
  passing_score?: number | null
  deleted: boolean
  client_id: string
  version: number
  created_at?: string | null
  last_modified?: string | null
}

// Individual question
export type QuestionBase = {
  id: number
  quiz_id: number
  question_type: QuestionType
  question_text: string
  options?: string[] | null
  hint?: string | null
  hint_penalty_points?: number | null
  source_citations?: SourceCitation[] | null
  points: number
  order_index: number
  tags?: string[] | null
  deleted: boolean
  client_id: string
  version: number
  created_at?: string | null
  last_modified?: string | null
}

export type QuestionPublic = QuestionBase

export type QuestionAdmin = QuestionBase & {
  correct_answer: AnswerValue
  explanation?: string | null
}

export type Question = QuestionAdmin

// Quiz attempt answer
export type QuizAnswer = {
  question_id: number
  user_answer: AnswerValue
  is_correct: boolean
  correct_answer?: AnswerValue
  explanation?: string | null
  hint_used?: boolean | null
  hint_penalty_points?: number | null
  source_citations?: SourceCitation[] | null
  points_awarded?: number | null
  time_spent_ms?: number | null
}

export type QuizAnswerInput = {
  question_id: number
  user_answer: AnswerValue
  hint_used?: boolean | null
  time_spent_ms?: number | null
}

// Quiz attempt/session
export type QuizAttempt = {
  id: number
  quiz_id: number
  started_at: string
  completed_at?: string | null
  score?: number | null
  total_possible: number
  time_spent_seconds?: number | null
  answers: QuizAnswer[]
  questions?: QuestionPublic[]
}

// Create types
export type QuizCreate = {
  name: string
  description?: string | null
  workspace_id?: string | null
  workspace_tag?: string | null
  media_id?: number | null
  source_bundle_json?: QuizGenerateSource[] | null
  time_limit_seconds?: number | null
  passing_score?: number | null
}

export type QuizUpdate = {
  name?: string | null
  description?: string | null
  workspace_id?: string | null
  workspace_tag?: string | null
  media_id?: number | null
  source_bundle_json?: QuizGenerateSource[] | null
  time_limit_seconds?: number | null
  passing_score?: number | null
  expected_version?: number | null
}

export type QuestionCreate = {
  question_type: QuestionType
  question_text: string
  options?: string[] | null
  correct_answer: AnswerValue
  explanation?: string | null
  hint?: string | null
  hint_penalty_points?: number | null
  source_citations?: SourceCitation[] | null
  points?: number
  order_index?: number
  tags?: string[] | null
}

export type QuestionUpdate = {
  question_type?: QuestionType | null
  question_text?: string | null
  options?: string[] | null
  correct_answer?: AnswerValue | null
  explanation?: string | null
  hint?: string | null
  hint_penalty_points?: number | null
  source_citations?: SourceCitation[] | null
  points?: number | null
  order_index?: number | null
  tags?: string[] | null
  expected_version?: number | null
}

// AI generation request
type QuizGenerateRequestBase = {
  num_questions?: number
  question_types?: QuestionType[]
  difficulty?: "easy" | "medium" | "hard" | "mixed"
  focus_topics?: string[]
  model?: string
  workspace_id?: string | null
  workspace_tag?: string | null
}

type QuizGenerateRequestWithMedia = QuizGenerateRequestBase & {
  media_id: number
  sources?: QuizGenerateSource[]
}

type QuizGenerateRequestWithSources = QuizGenerateRequestBase & {
  sources: QuizGenerateSource[]
  media_id?: number
}

export type QuizGenerateRequest = QuizGenerateRequestWithMedia | QuizGenerateRequestWithSources

export type QuizRemediationGenerateRequest = {
  attemptId: number
  questionIds: number[]
  num_questions?: number
  question_types?: QuestionType[]
  difficulty?: "easy" | "medium" | "hard" | "mixed"
  focus_topics?: string[]
  model?: string
  api_provider?: string
  workspace_tag?: string | null
}

export type QuizRemediationConversionSummary = {
  id: number
  attempt_id: number
  quiz_id: number
  question_id: number
  status: "active" | "superseded"
  orphaned: boolean
  superseded_count: number
  superseded_by_id?: number | null
  target_deck_id?: number | null
  target_deck_name_snapshot?: string | null
  flashcard_count: number
  flashcard_uuids_json: string[]
  source_ref_id?: string | null
  created_at?: string | null
  last_modified?: string | null
  client_id: string
  version: number
}

export type QuizRemediationConversionListResponse = {
  attempt_id: number
  items: QuizRemediationConversionSummary[]
  count: number
  superseded_count: number
}

export type QuizRemediationTargetDeck = {
  id: number
  name: string
}

export type QuizRemediationConvertRequest = {
  question_ids: number[]
  target_deck_id?: number | null
  create_deck_name?: string | null
  create_deck_review_prompt_side?: DeckReviewPromptSide
  create_deck_scheduler_type?: DeckSchedulerType | null
  create_deck_scheduler_settings?: DeckSchedulerSettingsEnvelope | null
  replace_active?: boolean
}

export type QuizRemediationConvertResult = {
  question_id: number
  status: "created" | "already_exists" | "superseded_and_created" | "failed"
  conversion?: QuizRemediationConversionSummary | null
  flashcard_uuids: string[]
  error?: string | null
}

export type QuizRemediationConvertResponse = {
  attempt_id: number
  quiz_id: number
  target_deck?: QuizRemediationTargetDeck | null
  results: QuizRemediationConvertResult[]
  created_flashcard_uuids: string[]
}

// List response types
export type QuizListResponse = {
  items: Quiz[]
  count: number
}

export type QuestionListResponse = {
  items: Array<QuestionPublic | QuestionAdmin>
  count: number
}

export type AttemptListResponse = {
  items: QuizAttempt[]
  count: number
}

// List params
export type QuizListParams = {
  media_id?: number | null
  q?: string | null
  workspace_id?: string | null
  workspace_tag?: string | null
  include_workspace_items?: boolean | null
  limit?: number
  offset?: number
}

export type QuestionListParams = {
  include_answers?: boolean
  q?: string | null
  limit?: number
  offset?: number
}

export type AttemptListParams = {
  quiz_id?: number | null
  limit?: number
  offset?: number
}

export type QuizGenerateResponse = {
  quiz: Quiz
  questions: QuestionAdmin[]
}

export type QuizImportQuestion = {
  question_type: QuestionType
  question_text: string
  options?: string[] | null
  correct_answer: AnswerValue
  explanation?: string | null
  hint?: string | null
  hint_penalty_points?: number | null
  source_citations?: SourceCitation[] | null
  points?: number
  order_index?: number
  tags?: string[] | null
}

export type QuizImportEntry = {
  quiz: QuizCreate
  questions: QuizImportQuestion[]
}

export type QuizImportRequest = {
  export_format?: string | null
  quizzes: QuizImportEntry[]
}

export type QuizImportItemResult = {
  source_index: number
  quiz_id: number
  imported_questions: number
  failed_questions: number
}

export type QuizImportError = {
  source_index: number
  quiz_name?: string | null
  question_index?: number | null
  error: string
}

export type QuizImportResponse = {
  imported_quizzes: number
  failed_quizzes: number
  imported_questions: number
  failed_questions: number
  items: QuizImportItemResult[]
  errors: QuizImportError[]
}

// --- Quizzes CRUD ---

export async function listQuizzes(params: QuizListParams = {}): Promise<QuizListResponse> {
  return await quizzesClient.list<QuizListResponse>({
    media_id: params.media_id,
    q: params.q,
    workspace_id: params.workspace_id,
    workspace_tag: params.workspace_tag,
    include_workspace_items: params.include_workspace_items ?? false,
    limit: params.limit,
    offset: params.offset
  })
}

export async function createQuiz(input: QuizCreate): Promise<Quiz> {
  return await quizzesClient.create<Quiz>(input)
}

export async function getQuiz(quizId: number): Promise<Quiz> {
  return await quizzesClient.get<Quiz>(quizId)
}

export async function updateQuiz(quizId: number, input: QuizUpdate): Promise<void> {
  await quizzesClient.update<void>(quizId, input)
}

export async function deleteQuiz(quizId: number, expectedVersion: number): Promise<void> {
  await quizzesClient.remove<void>(quizId, {
    expected_version: expectedVersion
  })
}

// --- Questions CRUD ---

export async function listQuestions(
  quizId: number,
  params: QuestionListParams = {}
): Promise<QuestionListResponse> {
  return await getQuestionsClient(quizId).list<QuestionListResponse>({
    include_answers: params.include_answers ? true : undefined,
    q: params.q,
    limit: params.limit,
    offset: params.offset
  })
}

export async function createQuestion(quizId: number, input: QuestionCreate): Promise<Question> {
  return await getQuestionsClient(quizId).create<Question>(input)
}

export async function updateQuestion(
  quizId: number,
  questionId: number,
  input: QuestionUpdate
): Promise<void> {
  await getQuestionsClient(quizId).update<void>(questionId, input)
}

export async function deleteQuestion(
  quizId: number,
  questionId: number,
  expectedVersion: number
): Promise<void> {
  await getQuestionsClient(quizId).remove<void>(questionId, {
    expected_version: expectedVersion
  })
}

// --- Quiz Attempts ---

export async function startAttempt(quizId: number): Promise<QuizAttempt> {
  return await getQuizAttemptsClient(quizId).create<QuizAttempt>({})
}

export async function submitAttempt(
  attemptId: number,
  answers: QuizAnswerInput[]
): Promise<QuizAttempt> {
  return await quizAttemptsClient.update<QuizAttempt>(attemptId, { answers })
}

export async function listAttempts(params: AttemptListParams = {}): Promise<AttemptListResponse> {
  return await quizAttemptsClient.list<AttemptListResponse>({
    quiz_id: params.quiz_id,
    limit: params.limit,
    offset: params.offset
  })
}

export async function getAttempt(
  attemptId: number,
  params?: {
    include_questions?: boolean
    include_answers?: boolean
  }
): Promise<QuizAttempt> {
  return await quizAttemptsClient.get<QuizAttempt>(attemptId, {
    include_questions: params?.include_questions ? true : undefined,
    include_answers: params?.include_answers ? true : undefined
  })
}

export async function listAttemptRemediationConversions(
  attemptId: number,
  options?: { signal?: AbortSignal }
): Promise<QuizRemediationConversionListResponse> {
  return await bgRequest<QuizRemediationConversionListResponse, AllowedPath, "GET">({
    path: `/api/v1/quizzes/attempts/${attemptId}/remediation-conversions` as AllowedPath,
    method: "GET",
    abortSignal: options?.signal
  })
}

export async function convertAttemptRemediationQuestions(
  attemptId: number,
  input: QuizRemediationConvertRequest,
  options?: { signal?: AbortSignal }
): Promise<QuizRemediationConvertResponse> {
  return await bgRequest<QuizRemediationConvertResponse, AllowedPath, "POST">({
    path: `/api/v1/quizzes/attempts/${attemptId}/remediation-conversions/convert` as AllowedPath,
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: input,
    abortSignal: options?.signal
  })
}

export async function getQuizAttemptQuestionAssistant(
  attemptId: number,
  questionId: number,
  options?: { signal?: AbortSignal }
): Promise<StudyAssistantContextResponse> {
  return await bgRequest<StudyAssistantContextResponse, AllowedPath, "GET">({
    path: `/api/v1/quizzes/attempts/${attemptId}/questions/${questionId}/assistant` as AllowedPath,
    method: "GET",
    abortSignal: options?.signal
  })
}

export async function respondQuizAttemptQuestionAssistant(
  attemptId: number,
  questionId: number,
  input: StudyAssistantRespondRequest,
  options?: { signal?: AbortSignal }
): Promise<StudyAssistantRespondResponse> {
  return await bgRequest<StudyAssistantRespondResponse, AllowedPath, "POST">({
    path: `/api/v1/quizzes/attempts/${attemptId}/questions/${questionId}/assistant/respond` as AllowedPath,
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: input,
    abortSignal: options?.signal
  })
}

// --- AI Generation ---

export async function generateQuiz(
  request: QuizGenerateRequest,
  options?: { signal?: AbortSignal; timeoutMs?: number }
): Promise<QuizGenerateResponse> {
  const timeoutMs =
    typeof options?.timeoutMs === "number" && options.timeoutMs > 0
      ? options.timeoutMs
      : QUIZ_GENERATION_TIMEOUT_MS

  return await bgRequest<QuizGenerateResponse, AllowedPath, "POST">({
    path: "/api/v1/quizzes/generate",
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: request,
    abortSignal: options?.signal,
    timeoutMs
  })
}

export function buildQuizAttemptQuestionSources(
  attemptId: number,
  questionIds: number[]
): QuizGenerateSource[] {
  return questionIds.map((questionId) => ({
    source_type: "quiz_attempt_question",
    source_id: `${attemptId}:${questionId}`
  }))
}

export async function generateRemediationQuiz(
  request: QuizRemediationGenerateRequest,
  options?: { signal?: AbortSignal; timeoutMs?: number }
): Promise<QuizGenerateResponse> {
  return await generateQuiz(
    {
      num_questions: request.num_questions,
      question_types: request.question_types,
      difficulty: request.difficulty,
      focus_topics: request.focus_topics,
      model: request.model,
      api_provider: request.api_provider,
      workspace_tag: request.workspace_tag,
      sources: buildQuizAttemptQuestionSources(request.attemptId, request.questionIds)
    },
    options
  )
}

export async function importQuizzesJson(
  request: QuizImportRequest,
  options?: { signal?: AbortSignal }
): Promise<QuizImportResponse> {
  return await bgRequest<QuizImportResponse, AllowedPath, "POST">({
    path: "/api/v1/quizzes/import/json",
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: request,
    abortSignal: options?.signal
  })
}

File: tldw_Server_API/app/api/v1/schemas/quizzes.py
from enum import Enum
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .flashcards import (
    DeckReviewPromptSide,
    DeckSchedulerSettingsEnvelope,
    DeckSchedulerType,
    _coerce_scheduler_settings_envelope,
)


class QuestionType(str, Enum):
    MULTIPLE_CHOICE = "multiple_choice"
    MULTI_SELECT = "multi_select"
    MATCHING = "matching"
    TRUE_FALSE = "true_false"
    FILL_BLANK = "fill_blank"


class QuizSourceType(str, Enum):
    MEDIA = "media"
    NOTE = "note"
    FLASHCARD_DECK = "flashcard_deck"
    FLASHCARD_CARD = "flashcard_card"
    QUIZ_ATTEMPT = "quiz_attempt"
    QUIZ_ATTEMPT_QUESTION = "quiz_attempt_question"


AnswerValue = int | str | list[int] | dict[str, str]


class SourceCitation(BaseModel):
    source_type: Optional[QuizSourceType] = None
    source_id: Optional[str] = Field(None, min_length=1)
    label: Optional[str] = None
    quote: Optional[str] = None
    media_id: Optional[int] = Field(None, ge=1)
    chunk_id: Optional[str] = None
    timestamp_seconds: Optional[float] = Field(None, ge=0)
    source_url: Optional[str] = None


class QuizGenerateSource(BaseModel):
    source_type: QuizSourceType
    source_id: str = Field(..., min_length=1)


class QuizCreate(BaseModel):
    name: str = Field(..., description="Quiz name")
    description: Optional[str] = Field(None, description="Optional quiz description")
    workspace_tag: Optional[str] = Field(None, description="Optional workspace tag (e.g., 'workspace:<slug-or-id>')")
    workspace_id: Optional[str] = Field(None, description="Canonical owning workspace ID; null means general scope")
    media_id: Optional[int] = Field(None, description="Source media ID for AI-generated quizzes")
    source_bundle_json: Optional[list[QuizGenerateSource]] = Field(
        None, description="Optional canonical mixed-source bundle used to generate this quiz"
    )
    time_limit_seconds: Optional[int] = Field(None, ge=1, description="Optional time limit in seconds")
    passing_score: Optional[int] = Field(None, ge=0, le=100, description="Passing score percentage")


class QuizUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: Optional[str] = None
    description: Optional[str] = None
    workspace_id: Optional[str] = None
    workspace_tag: Optional[str] = None
    media_id: Optional[int] = None
    source_bundle_json: Optional[list[QuizGenerateSource]] = None
    time_limit_seconds: Optional[int] = Field(None, ge=1)
    passing_score: Optional[int] = Field(None, ge=0, le=100)
    expected_version: Optional[int] = None


class QuizResponse(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    workspace_tag: Optional[str] = None
    workspace_id: Optional[str] = None
    media_id: Optional[int] = None
    source_bundle_json: Optional[list[QuizGenerateSource]] = None
    total_questions: int
    time_limit_seconds: Optional[int] = None
    passing_score: Optional[int] = None
    deleted: bool
    client_id: str
    version: int
    created_at: Optional[str] = None
    last_modified: Optional[str] = None


class QuizListResponse(BaseModel):
    items: list[QuizResponse]
    count: int


class QuestionCreate(BaseModel):
    question_type: QuestionType
    question_text: str
    options: Optional[list[str]] = Field(None, description="Multiple choice options")
    correct_answer: AnswerValue
    explanation: Optional[str] = None
    hint: Optional[str] = None
    hint_penalty_points: int = Field(0, ge=0)
    source_citations: Optional[list[SourceCitation]] = None
    points: int = Field(1, ge=0)
    order_index: int = 0
    tags: Optional[list[str]] = None


class QuestionUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    question_type: Optional[QuestionType] = None
    question_text: Optional[str] = None
    options: Optional[list[str]] = None
    correct_answer: Optional[AnswerValue] = None
    explanation: Optional[str] = None
    hint: Optional[str] = None
    hint_penalty_points: Optional[int] = Field(None, ge=0)
    source_citations: Optional[list[SourceCitation]] = None
    points: Optional[int] = Field(None, ge=0)
    order_index: Optional[int] = None
    tags: Optional[list[str]] = None
    expected_version: Optional[int] = None


class QuestionPublicResponse(BaseModel):
    id: int
    quiz_id: int
    question_type: QuestionType
    question_text: str
    options: Optional[list[str]] = None
    hint: Optional[str] = None
    hint_penalty_points: int = Field(0, ge=0)
    source_citations: Optional[list[SourceCitation]] = None
    points: int
    order_index: int
    tags: Optional[list[str]] = None
    deleted: bool
    client_id: str
    version: int
    created_at: Optional[str] = None
    last_modified: Optional[str] = None


class QuestionAdminResponse(QuestionPublicResponse):
    correct_answer: Optional[AnswerValue] = None
    explanation: Optional[str] = None


class QuestionListResponse(BaseModel):
    items: list[QuestionPublicResponse | QuestionAdminResponse]
    count: int


class QuizAnswerInput(BaseModel):
    question_id: int
    user_answer: AnswerValue
    hint_used: Optional[bool] = None
    time_spent_ms: Optional[int] = None


class AttemptSubmitRequest(BaseModel):
    answers: list[QuizAnswerInput]


class AttemptAnswer(BaseModel):
    question_id: int
    user_answer: AnswerValue
    is_correct: bool
    correct_answer: Optional[AnswerValue] = None
    explanation: Optional[str] = None
    hint_used: Optional[bool] = None
    hint_penalty_points: Optional[int] = Field(None, ge=0)
    source_citations: Optional[list[SourceCitation]] = None
    points_awarded: Optional[int] = None
    time_spent_ms: Optional[int] = None


class AttemptResponse(BaseModel):
    id: int
    quiz_id: int
    started_at: str
    completed_at: Optional[str] = None
    score: Optional[int] = None
    total_possible: int
    time_spent_seconds: Optional[int] = None
    answers: list[AttemptAnswer] = Field(default_factory=list)
    questions: Optional[list[QuestionPublicResponse]] = None


class AttemptListResponse(BaseModel):
    items: list[AttemptResponse]
    count: int


class QuizRemediationConversionSummary(BaseModel):
    id: int
    attempt_id: int
    quiz_id: int
    question_id: int
    status: Literal["active", "superseded"]
    orphaned: bool = False
    superseded_count: int = Field(0, ge=0)
    superseded_by_id: Optional[int] = None
    target_deck_id: Optional[int] = None
    target_deck_name_snapshot: Optional[str] = None
    flashcard_count: int = Field(0, ge=0)
    flashcard_uuids_json: list[str] = Field(default_factory=list)
    source_ref_id: Optional[str] = None
    created_at: Optional[str] = None
    last_modified: Optional[str] = None
    client_id: str
    version: int


class QuizRemediationConversionListResponse(BaseModel):
    attempt_id: int
    items: list[QuizRemediationConversionSummary] = Field(default_factory=list)
    count: int = Field(..., ge=0)
    superseded_count: int = Field(..., ge=0)


class QuizRemediationTargetDeck(BaseModel):
    id: int
    name: str


class QuizRemediationConvertRequest(BaseModel):
    question_ids: list[int] = Field(..., min_length=1)
    target_deck_id: Optional[int] = Field(None, ge=1)
    create_deck_name: Optional[str] = None
    create_deck_review_prompt_side: DeckReviewPromptSide = "front"
    create_deck_scheduler_type: Optional[DeckSchedulerType] = None
    create_deck_scheduler_settings: Optional[DeckSchedulerSettingsEnvelope] = None
    replace_active: bool = False

    @model_validator(mode="after")
    def _reject_explicit_null_create_deck_review_prompt_side(self) -> "QuizRemediationConvertRequest":
        if (
            "create_deck_review_prompt_side" in self.model_fields_set
            and self.create_deck_review_prompt_side is None
        ):
            raise ValueError("create_deck_review_prompt_side cannot be null")
        return self

    @model_validator(mode="before")
    def normalize_scheduler_settings(cls, data):
        if not isinstance(data, dict):
            return data
        if "create_deck_scheduler_settings" in data:
            data["create_deck_scheduler_settings"] = _coerce_scheduler_settings_envelope(
                data.get("create_deck_scheduler_settings")
            )
        return data

    @model_validator(mode="after")
    def validate_deck_target(self) -> "QuizRemediationConvertRequest":
        has_target_deck = self.target_deck_id is not None
        has_create_deck = bool((self.create_deck_name or "").strip())
        if has_target_deck == has_create_deck:
            raise ValueError("Provide exactly one of target_deck_id or create_deck_name")
        if "create_deck_review_prompt_side" in self.model_fields_set and not has_create_deck:
            raise ValueError("create_deck review orientation requires create_deck_name")
        if (self.create_deck_scheduler_settings is not None or self.create_deck_scheduler_type is not None) and not has_create_deck:
            raise ValueError("create_deck scheduler options require create_deck_name")
        return self


class QuizRemediationConvertResult(BaseModel):
    question_id: int
    status: Literal["created", "already_exists", "superseded_and_created", "failed"]
    conversion: Optional[QuizRemediationConversionSummary] = None
    flashcard_uuids: list[str] = Field(default_factory=list)
    error: Optional[str] = None


class QuizRemediationConvertResponse(BaseModel):
    attempt_id: int
    quiz_id: int
    target_deck: Optional[QuizRemediationTargetDeck] = None
    results: list[QuizRemediationConvertResult] = Field(default_factory=list)
    created_flashcard_uuids: list[str] = Field(default_factory=list)


class QuizGenerateRequest(BaseModel):
    media_id: Optional[int] = Field(None, ge=1)
    sources: Optional[list[QuizGenerateSource]] = Field(None, min_length=1)
    num_questions: int = Field(10, ge=1, le=100)
    question_types: Optional[list[QuestionType]] = None
    difficulty: str = Field("mixed", description="easy, medium, hard, mixed")
    focus_topics: Optional[list[str]] = None
    model: Optional[str] = None
    api_provider: Optional[str] = None
    workspace_id: Optional[str] = Field(None, description="Canonical owning workspace ID; null means general scope")
    workspace_tag: Optional[str] = Field(None, description="Optional workspace tag (e.g., 'workspace:<slug-or-id>')")

    @model_validator(mode="after")
    def validate_media_id_or_sources(self) -> "QuizGenerateRequest":
        if self.media_id is None and not self.sources:
            raise ValueError("Either media_id or sources must be provided")
        return self


class QuizGenerateResponse(BaseModel):
    quiz: QuizResponse
    questions: list[QuestionAdminResponse]


class QuizImportQuestion(BaseModel):
    question_type: QuestionType
    question_text: str
    options: Optional[list[str]] = None
    correct_answer: AnswerValue
    explanation: Optional[str] = None
    hint: Optional[str] = None
    hint_penalty_points: int = Field(0, ge=0)
    source_citations: Optional[list[SourceCitation]] = None
    points: int = Field(1, ge=0)
    order_index: int = Field(0, ge=0)
    tags: Optional[list[str]] = None


class QuizImportQuiz(BaseModel):
    name: str
    description: Optional[str] = None
    workspace_id: Optional[str] = None
    workspace_tag: Optional[str] = None
    media_id: Optional[int] = None
    source_bundle_json: Optional[list[QuizGenerateSource]] = None
    time_limit_seconds: Optional[int] = Field(None, ge=1)
    passing_score: Optional[int] = Field(None, ge=0, le=100)


class QuizImportEntry(BaseModel):
    quiz: QuizImportQuiz
    questions: list[QuizImportQuestion] = Field(default_factory=list)


class QuizImportRequest(BaseModel):
    export_format: Optional[str] = Field(
        default=None,
        description="Expected export format marker, e.g. tldw.quiz.export.v1",
    )
    quizzes: list[QuizImportEntry]


class QuizImportItemResult(BaseModel):
    source_index: int = Field(..., ge=0, description="Index of the input quiz entry")
    quiz_id: int
    imported_questions: int = Field(..., ge=0)
    failed_questions: int = Field(..., ge=0)


class QuizImportError(BaseModel):
    source_index: int = Field(..., ge=0)
    quiz_name: Optional[str] = None
    question_index: Optional[int] = Field(default=None, ge=0)
    error: str


class QuizImportResponse(BaseModel):
    imported_quizzes: int = Field(..., ge=0)
    failed_quizzes: int = Field(..., ge=0)
    imported_questions: int = Field(..., ge=0)
    failed_questions: int = Field(..., ge=0)
    items: list[QuizImportItemResult] = Field(default_factory=list)
    errors: list[QuizImportError] = Field(default_factory=list)
